File: src/main/java/kr/co/lunatalk/domain/auth/service/AuthService.java
package kr.co.lunatalk.domain.auth.service;

import kr.co.lunatalk.domain.auth.dto.RefreshTokenDto;
import kr.co.lunatalk.domain.auth.dto.request.LoginRequest;
import kr.co.lunatalk.domain.auth.dto.request.RefreshTokenRequest;
import kr.co.lunatalk.domain.auth.dto.response.AuthTokenResponse;
import kr.co.lunatalk.domain.auth.dto.response.TokenResponse;
import kr.co.lunatalk.domain.member.domain.Member;
import kr.co.lunatalk.domain.member.domain.MemberStatus;
import kr.co.lunatalk.domain.member.domain.Profile;
import kr.co.lunatalk.domain.member.dto.request.CreateMemberRequest;
import kr.co.lunatalk.domain.member.repository.MemberRepository;
import kr.co.lunatalk.global.exception.CustomException;
import kr.co.lunatalk.global.exception.ErrorCode;
import kr.co.lunatalk.global.security.JwtTokenProvider;
import kr.co.lunatalk.global.util.MemberUtil;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class AuthService {
	private final MemberRepository memberRepository;
	private final PasswordEncoder passwordEncoder;
	private final JwtTokenProvider jwtTokenProvider;
	private final MemberUtil memberUtil;


	public AuthTokenResponse registerMember(CreateMemberRequest request) {

		Optional<Member> existsMember = memberRepository.findByUsername(request.username());
		if (existsMember.isPresent()) {
			throw new CustomException(ErrorCode.MEMBER_EXISTS);
		}

		Member member = Member.of(request.username(), encodePassword(request.password()), Profile.of("", ""));
		memberRepository.save(member);

		TokenResponse token = getTokenResponse(member);

		return AuthTokenResponse.from(token);
	}

	public AuthTokenResponse loginMember(LoginRequest request) {
		Optional<Member> findMember = memberRepository.findByUsername(request.username());

		if (findMember.isEmpty()) {
			throw new CustomException(ErrorCode.MEMBER_NOT_FOUND);
		}

		Member member = findMember.get();

		if (member.getStatus() == MemberStatus.DELETE) {
			throw new CustomException(ErrorCode.MEMBER_NOT_FOUND);
		}

		Boolean isMatching = matchingPassword(request.password(), member.getPassword());

		if(!isMatching) {
			throw new CustomException(ErrorCode.AUTH_UNAUTHORIZED);
		}

		return AuthTokenResponse.from(getTokenResponse(member));
	}

	public void withdraw() {
		Member currentMember = memberUtil.getCurrentMember();

		if(currentMember.getStatus() == MemberStatus.DELETE) {
			throw new CustomException(ErrorCode.MEMBER_ALREADY_DELETED);
		}
		currentMember.withdrawal();
		memberRepository.deleteById(currentMember.getId());
		jwtTokenProvider.deleteRefreshTokenFromRedis(currentMember.getId());
	}

	private TokenResponse getTokenResponse(Member member) {
		return jwtTokenProvider.generateTokenPair(member.getId(), member.getRole());
	}

	@Transactional(readOnly = true)
	public AuthTokenResponse reissueTokenPair(RefreshTokenRequest request) {
		RefreshTokenDto refreshTokenDto = jwtTokenProvider.retrieveRefreshToken(request.refreshToken());

		if(refreshTokenDto == null) {
			throw new CustomException(ErrorCode.AUTH_TOKEN_EXPIRED);
		}

		Optional<Member> findMember = memberRepository.findById(refreshTokenDto.memberId());

		if (findMember.isEmpty()) {
			throw new CustomException(ErrorCode.MEMBER_NOT_FOUND);
		}

		Member member = findMember.get();

		return AuthTokenResponse.from(getTokenResponse(member));
	}

	private String encodePassword(String password) {
		return passwordEncoder.encode(password);
	}

	private Boolean matchingPassword(String password, String encodedPassword) {
		return passwordEncoder.matches(password, encodedPassword);
	}
}

File: lunatalk-domain/src/main/kotlin/kr/co/lunatalk/domain/auth/service/AuthService.kt
package kr.co.lunatalk.domain.auth.service

import kr.co.lunatalk.domain.auth.dto.request.LoginRequest
import kr.co.lunatalk.domain.auth.dto.request.RefreshTokenRequest
import kr.co.lunatalk.domain.auth.dto.response.AuthTokenResponse
import kr.co.lunatalk.domain.auth.dto.response.TokenResponse
import kr.co.lunatalk.domain.member.domain.Member
import kr.co.lunatalk.domain.member.domain.MemberRole
import kr.co.lunatalk.domain.member.domain.MemberStatus
import kr.co.lunatalk.domain.member.domain.Profile
import kr.co.lunatalk.domain.member.dto.request.CreateMemberRequest
import kr.co.lunatalk.domain.member.repository.MemberRepository
import kr.co.lunatalk.global.exception.CustomException
import kr.co.lunatalk.global.exception.ErrorCode
import kr.co.lunatalk.global.security.JwtTokenProvider
import kr.co.lunatalk.global.util.MemberUtil
import org.slf4j.LoggerFactory
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class AuthService(
    private val memberRepository: MemberRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtTokenProvider: JwtTokenProvider,
    private val memberUtil: MemberUtil,
) {

    private val log = LoggerFactory.getLogger(AuthService::class.java)

    fun registerMember(request: CreateMemberRequest): AuthTokenResponse {
        val existsMember = memberRepository.findByUsername(request.username)
        if (existsMember.isPresent) {
            throw CustomException(ErrorCode.MEMBER_EXISTS)
        }

        val member = Member.createMember(
            request.username,
            encodePassword(request.password),
            Profile.of("", ""),
            request.phone,
            request.email,
        )
        memberRepository.save(member)

        val token = getTokenResponse(member)
        return AuthTokenResponse.from(token)
    }

    fun loginMember(request: LoginRequest): AuthTokenResponse {
        val member = findMemberByUsername(request.username)

        if (member.status == MemberStatus.DELETE) {
            throw CustomException(ErrorCode.MEMBER_NOT_FOUND)
        }

        return passwordMatchingAndTokenPair(request, member)
    }

    fun loginAdmin(request: LoginRequest): AuthTokenResponse {
        val member = findMemberByUsername(request.username)

        if (member.role != MemberRole.ADMIN) {
            throw CustomException(ErrorCode.AUTH_UNAUTHORIZED)
        }

        return passwordMatchingAndTokenPair(request, member)
    }

    private fun passwordMatchingAndTokenPair(request: LoginRequest, member: Member): AuthTokenResponse {
        val isMatching = matchingPassword(request.password, member.password ?: "")

        if (!isMatching) {
            throw CustomException(ErrorCode.AUTH_UNAUTHORIZED)
        }

        return AuthTokenResponse.from(getTokenResponse(member))
    }

    private fun findMemberByUsername(username: String): Member =
        memberRepository.findByUsername(username).orElseThrow {
            CustomException(ErrorCode.MEMBER_NOT_FOUND)
        }

    fun withdraw() {
        val currentMember = memberUtil.currentMember

        if (currentMember.status == MemberStatus.DELETE) {
            throw CustomException(ErrorCode.MEMBER_ALREADY_DELETED)
        }
        currentMember.withdrawal()
        memberRepository.deleteById(currentMember.id!!)
        jwtTokenProvider.deleteRefreshTokenFromRedis(currentMember.id!!)
    }

    private fun getTokenResponse(member: Member): TokenResponse =
        jwtTokenProvider.generateTokenPair(member.id!!, member.role)

    @Transactional(readOnly = true)
    fun reissueTokenPair(request: RefreshTokenRequest): AuthTokenResponse {
        val refreshTokenDto = jwtTokenProvider.retrieveRefreshToken(request.refreshToken)
            ?: throw CustomException(ErrorCode.AUTH_REFRESH_TOKEN_EXPIRED)

        val member = memberRepository.findById(refreshTokenDto.memberId).orElseThrow {
            CustomException(ErrorCode.MEMBER_NOT_FOUND)
        }

        return AuthTokenResponse.from(getTokenResponse(member))
    }

    private fun encodePassword(password: String): String =
        passwordEncoder.encode(password)!!

    private fun matchingPassword(password: String, encodedPassword: String): Boolean =
        passwordEncoder.matches(password, encodedPassword)
}
